flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/DefaultLeaderElectionServiceTest.java

```diff
@@ -1246,6 +1200,62 @@ private void testNonBlockingCall(
         testInstance.close();
     }
 
+    /**
+     * This test is used to verify FLINK-36451 where we observed concurrent nested locks being
+     * acquired from the {@link LeaderContender} and from the {@link DefaultLeaderElectionService}.
+     */
+    @Test
+    void testNestedDeadlockInLeadershipConfirmation() throws Exception {
```

Comment:
That's the test that I added to verify the findings. I kept it in a separate commit to prove that the test runs into a deadlock before applying the changes.